Implement Robust comp time versioning of binaries

Diogenesoftoronto opened the pull request:
This PR is now fixed. It creates but does not actually implement the new version command via ffcli. This will be tackled in a follow up pr along with a new command server command. This will give enduro a similiar structure to our other projects. 

Refs #520, #518

Diogenesoftoronto committed:
Implement Robust comp time versioning of binaries

djjuhasz commented on internal/version/version.go:
```diff
@@ -1,32 +1,68 @@
 	"runtime/debug"
+	"strings"
+
+	enduroroot "github.com/artefactual-sdps/enduro"
```
Did you mean to import "github.com/artefactual-sdps/enduroroot" here?  I think importing github.com/artefactual-sdps/enduro will cause a recursive import?

Diogenesoftoronto replied:
This doesn't cause circular imports, I am pretty sure I would get circular import errors but I don't. Also, I modelled it the same way as sdps-api which also imports the root.

djjuhasz commented on Dockerfile:
```diff
@@ -12,7 +12,7 @@ COPY --link . .
 RUN --mount=type=cache,target=/go/pkg/mod \
 	--mount=type=cache,target=/root/.cache/go-build \
-	go build -o /out/enduro .
+	go build -o /out/enduro ./cmd/enduro-ctl
```
I think ideally we want to build the binaries with the version flags, e.g. 
```
    go build -trimpath -o /out/enduro \
    -ldflags="-X '${VERSION_PATH}.Long=${VERSION_LONG}' -X '${VERSION_PATH}.Short=${VERSION_SHORT}' -X '${VERSION_PATH}.GitCommit=${VERSION_GIT_HASH}'" \
    ./cmd/enduro-ctl
```

This would require some more work to get the version info from `build_dist.sh shellvars`.

Diogenesoftoronto replied:
That is a good point I will work on that

djjuhasz commented on internal/version/version.go:
```diff
@@ -1,32 +1,68 @@
+}
+
+func Info(appName string) string {
+	return fmt.Sprintf("%s version %s",
```
This seems short enough to be a one-liner.

Diogenesoftoronto replied:
sensible

Diogenesoftoronto committed:
more troubleshooting

djjuhasz committed:
Copy version files to Docker image

Copy the necessary version files when doing the Docker image build.

Diogenesoftoronto committed:
Add version arguments

Diogenesoftoronto committed:
Create a root command and package

Add a version command to replace the version flag.
Add the viper configuration to the ffcli command for futher flexibility.

Diogenesoftoronto committed:
Make viper configuration compatible with ffcli

Diogenesoftoronto committed:
Replace the api service name in the commands

Diogenesoftoronto committed:
Use the rootCommand's configuration struct

sevein commented on go.work:
```diff
@@ -2,5 +2,7 @@ go 1.20
 	.
+
 	./hack/pulumi
+
```
These blank lines you're adding I guess can be removed.

sevein commented on Dockerfile:
```diff
@@ -3,16 +3,24 @@
-	go build -o /out/enduro .
+    go build -trimpath -o /out/enduro \
+    -ldflags="-X '${VERSION_PATH}.Long=${VERSION_LONG}' -X '${VERSION_PATH}.Short=${VERSION_SHORT}' -X '${VERSION_PATH}.GitCommit=${VERSION_GIT_HASH}'" \
+    ./cmd/enduro-ctl
```
You're not using tabs here.

sevein reviewed (commented):
@Diogenesoftoronto, would it be too hard at this point to address one new pull request addressing #520 only? Once merged, we could go back to this pull request to focus on the use of ffcli and subcommands.

Diogenesoftoronto commented:
> @Diogenesoftoronto, would it be too hard at this point to address one new pull request addressing #520 only? Once merged, we could go back to this pull request to focus on the use of ffcli and subcommands.

It does seem like more work than it is worth at this point. I am already working on stuff that is based on this pr

Diogenesoftoronto committed:
Fix spacing and remove unnecessary imports

sevein commented on .bingo/Variables.mk:
```diff
@@ -1,4 +1,4 @@
-# Auto generated binary variables helper managed by https://github.com/bwplotka/bingo v0.8. DO NOT EDIT.
+# Auto generated binary variables helper managed by https://github.com/bwplotka/bingo v0.7. DO NOT EDIT.
```
Some changes under the .bingo directory that I don't think are relevant to this pull request.

Diogenesoftoronto replied:
will remove thanks

sevein commented on CHANGELOG.md:
```diff
@@ -0,0 +1,9 @@
+# Changelog
+
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog], and this project adheres to [Semantic
+Versioning].
+
+[keep a changelog]: https://keepachangelog.com/en/1.0.0/
+[semantic versioning]: https://semver.org/spec/v2.0.0.html
```
A changelog is cool but isn't that outside the scope of this pull request?

Diogenesoftoronto replied:
eh not really this seems to be relevant to version stuff it fits well enough and its useful.

Diogenesoftoronto replied:
I should add that in the final commit message and just change the PR Title to something like improve version project version management.

sevein replied:
Or you can keep the scope of this PR small and focused on the original issue: version stamping. We're the authors of many projects where the log of changes is communicated differently, I thought it'd be of your interest to consider other practices followed at Artefactual.

sevein commented on Dockerfile:
```diff
@@ -3,16 +3,24 @@
 ARG TARGET=enduro
 
 FROM golang:1.20.5-alpine AS build-go
+LABEL maintainer="Artefactual Systems Inc <info@artefactual.com>"
```
Question: is this label inherited by other images based on `build-go`?

sevein commented on Dockerfile:
```diff
@@ -3,16 +3,24 @@
 RUN --mount=type=cache,target=/go/pkg/mod \
 	--mount=type=cache,target=/root/.cache/go-build \
-	go build -o /out/enduro .
+    go build -trimpath -o /out/enduro \
```
I've noticed that you're not passing the new version flags to enduro-a3m-worker, why is that?

Diogenesoftoronto replied:
I should definitely do that nice catch

sevein commented on internal/rootcmd/root.go:
```diff
@@ -0,0 +1,37 @@
+	Verbose bool
+}
+
+func New() (*ffcli.Command, *Config) {
+	var cfg Config
+
+	fs := flag.NewFlagSet("enduro-ctl", flag.ExitOnError)
+	cfg.RegisterFlags(fs)
+
+	return &ffcli.Command{
+		Name:       "enduro-ctl",
+		ShortUsage: "enduro-ctl [flags] <subcommand> [flags] [<arg>...]",
+		FlagSet:    fs,
+		Exec:       cfg.Exec,
+	}, &cfg
+}
```
Are you using this constructor? Looks unused at the moment.

Diogenesoftoronto replied:
Yup it is unused

sevein reviewed (changes requested):
> It does seem like more work than it is worth at this point.

The way I see it the versioning part of this work is well defined and workign smothly - could be merged right away and we'd be making progress. The subcommands, on the other hand, I believe we need to explore a bit more. Subcommands work nicely for command-line interfaces like tctl, gh or kubectl that group numerous tasks. But I'm wondering if this is the best pattern for our services, e.g. enduro, enduro-a3m-worker, a possible enduro-storage, enduro-gateway, etc. A single binary with everything inside is going to take longer to build, have you considered that?

sevein commented on Dockerfile:
```diff
@@ -3,16 +3,24 @@
 
 FROM golang:1.20.5-alpine AS build-go
+LABEL maintainer="Artefactual Systems Inc <info@artefactual.com>"
+USER root
```
Was this added for some reason? What is it fixing?

sevein commented:
Superseded by https://github.com/artefactual-sdps/enduro/pull/623.
